src/nspawn/nspawn.c

```diff
@@ -2309,7 +2309,17 @@ static int mount_device(const char *what, const char *where, const char *directo
                 return -EOPNOTSUPP;
         }
 
-        return mount_verbose(LOG_ERR, what, p, fstype, MS_NODEV|(rw ? 0 : MS_RDONLY), NULL);
+        /* If this is a loopback device then let's mount the image with discard, so that the underlying file remains
+         * sparse when possible. */
+        if (STR_IN_SET(fstype, "btrfs", "ext4", "vfat", "xfs")) {
+                const char *l;
+
+                l = path_startswith(what, "/dev");
+                if (l && startswith(l, "loop"))
+                        options = "discard";
+        }
```

Comment:
So, what's the other option? checking stat() and the major/minor? So far we tried to avoid hardcoding major/minors, so that the kernel can assign them entirely dynamically as it likes (it does that only for some subsystems so far, but we really shouldn't hardcode for which one it does that and for which one it doesn't, as that might change...). 

So far the preferable way to identify the type of a device was either by the subsystem string (which doesn't work here, it's always "block") or by prefix checking of the device name, which is what we do here...

Yes, doing such a prefix name check isn't pretty, but I really see no better way to handle this. All the other options I see are much worse...